spring-boot/admin/src/main/java/eu/coatrack/admin/service/OAuthUserDetailsService.java

```diff
@@ -0,0 +1,106 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import eu.coatrack.config.github.GithubEmail;
+import lombok.AllArgsConstructor;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClientService;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestTemplate;
+
+import java.util.List;
+
+@Service
+@AllArgsConstructor
+public class OAuthUserDetailsService {
+
+    private static final String GITHUB_API_EMAIL = "https://api.github.com/user/emails";
+
+    private final OAuth2AuthorizedClientService clientService;
+    private final RestTemplate restTemplate;
+
+    private OAuth2User getLoggedInUser() {
+        return (OAuth2User) SecurityContextHolder
+                .getContext()
+                .getAuthentication()
+                .getPrincipal();
+    }
+
+    public String getLoginNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("login");
+    }
+
+    public String getNameFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("name");
+    }
+
+    public String getCompanyFromLoggedInUser() {
+        return getLoggedInUser().getAttribute("company");
+    }
+
+    public String getEmailFromLoggedInUser() throws JsonProcessingException {
+        String email = getLoggedInUser().getAttribute("email");
+        if (email == null || email.isEmpty()) {
+            email = getPrimaryEmailFromLoggedInUser();
+        }
+        return email;
+    }
+
+    private List<GithubEmail> getEmailsListFromGithub() {
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Authorization", "token " + getTokenFromLoggedInUser());
+        HttpEntity<String> githubRequest = new HttpEntity(headers);
+
+        ResponseEntity<List<GithubEmail>> emailsListFromGithub = restTemplate.exchange(GITHUB_API_EMAIL, HttpMethod.GET,
+                githubRequest, new ParameterizedTypeReference<List<GithubEmail>>() {});
+        return emailsListFromGithub.getBody();
+    }
+
+    private String getPrimaryEmailFromLoggedInUser() {
```

Comment:
A subfunction should be placed right below the super-function which calls the sub-function for the first time. The function getPrimaryEmailFromLoggedInUser() should be placed right below getEmailFromLoggedInUser() which calls this it as sub-method. This improves the reading flow, as the developer does not have to search for long, if he is interested in the details of getPrimaryEmailFromLoggedInUser().

Reply:
Done